lib/navigation/preprocess_imu.cpp

```diff
@@ -2,55 +2,99 @@
 
 namespace hyped::navigation {
 
-ImuPreprocessor::ImuPreprocessor()
+ImuPreprocessor::ImuPreprocessor(core::ILogger &logger)
+    : logger_(logger),
+      num_outliers_per_imu_({0, 0, 0, 0}),
+      are_imus_reliable_({true, true, true, true}),
+      num_reliable_accelerometers_(core::kNumImus)
 {
-  // TODOLater: implement
 }
 
-core::ImuData ImuPreprocessor::processData(const core::RawImuData raw_imu_data)
+std::optional<core::ImuData> ImuPreprocessor::processData(const core::RawImuData raw_imu_data)
 {
-  /*
-  TODOLater: implement
-  basic plan:
-  - call detectOutliers. Return of that function is return of this function
-  - call checkReliable
+  core::ImuData imu_data;
+  // TODOLater : check on direction of travel
+  core::Float magnitude;
+  for (std::size_t i = 0; i < core::kNumImus; ++i) {
+    magnitude = 0;
+    for (std::size_t j = 0; j < 3; ++j) {
+      magnitude += std::pow(raw_imu_data.at(i).at(j), 2);
+    }
+    imu_data.at(i) = std::sqrt(magnitude);
+  }
+  const core::ImuData accelerometer_data = detectOutliers(imu_data);
+  SensorChecks sensorcheck               = checkReliable();
 
-  */
-  return {0, 0, 0, 0};
+  if (sensorcheck == SensorChecks::kUnacceptable) {
+    return std::nullopt;
```

Comment:
Are we not wanting the logger to log a failure here?

Reply:
The logger logs the failure before sensor checks changes to kUnacceptable in the checkReliable function